### PR TITLE
Bulk move and swap artifacts rev2

### PR DESCRIPTION
**Pull request by SoundSSGood**
Implemented bulk artifact swap and bulk artifact move. And some refactoring. Thanks to @AgostonSzepessy  for the idea and the first step of implementation. 
Single artifact moving is not changed (May be changed to bulk in the future).
End of battle artifacts transfer is not changed (out of scope this PR).

### CODE REVIEW

**Comment by nullkiller** [comment]
Does it work on my map with composits? When swapping angelic allians and armor of damned for instance?

File: lib/CArtHandler.cpp [hunk 1]
```diff
@@ -1457,4 +1457,82 @@ void CArtifactSet::serializeJsonSlot(JsonSerializeFormat & handler, const Artifa
...
+	}
+}
+
+ArtBearer::ArtBearer CArtifactFittingSet::bearerType() const
+{
+	return this->Bearer;
+}
+
+DLL_LINKAGE ArtifactPosition ArtifactUtils::getArtifactDstPosition(const CArtifactInstance * artifact,
+	const CArtifactSet * target, ArtBearer::ArtBearer barer)
```
Comment (Nordsoft91) [inline lib/CArtHandler.cpp:1494]: artifact and target are never nullptr? you can use reference instead of pointers

**Review by Nordsoft91 — CHANGES_REQUESTED** [review]

**Comment by Nordsoft91** [comment]
@SoundSSGood does it fix the https://github.com/vcmi/vcmi/issues/840?

File: client/windows/GUIClasses.cpp [hunk 2]
```diff
@@ -955,56 +911,13 @@ std::vector<HeroArtifact> CExchangeController::moveCompositeArtsToBackpack()
...
-}
-
 std::function<void()> CExchangeController::onSwapArtifacts()
 {
 	return [&]()
 	{
 		GsThread::run([=]
 		{
-			// it is not possible directly exchange composite artifacts like Angelic Alliance and Armor of Damned
-			auto compositeArtLocations = moveCompositeArtsToBackpack();
```
[the remark this answers is not in the document]
Comment (SoundSSGood) [reply client/windows/GUIClasses.cpp:993]: yeah it can be removed

File: lib/CArtHandler.cpp (hunk 1 above)
Comment (nullkiller) [reply lib/CArtHandler.cpp:1494]: barer -> bearer

**Comment by nullkiller** [comment]
> @SoundSSGood does it fix the #840?

It looks unrelated. Better to fix in a separate PR

File: client/CPlayerInterface.cpp [hunk 3]
```diff
@@ -2591,6 +2591,11 @@ void CPlayerInterface::artifactMoved(const ArtifactLocation &src, const Artifact
 		if (artWin)
 			artWin->artifactMoved(src, dst);
 	}
+	GH.objsToBlit.back()->redraw();
```
[the remark this answers is not in the document]
Comment (SoundSSGood) [reply client/CPlayerInterface.cpp:2594]: According to the game logic, the vector will contain at least one value. But the CGuiHandler::simpleRedraw() method contains such a check. I will do the same.

**Comment by SoundSSGood** [comment]
> @SoundSSGood does it fix the #840?

shouldn't

File: CCallback.cpp [hunk 4]
```diff
@@ -181,6 +181,13 @@ bool CCallback::assembleArtifacts (const CGHeroInstance * hero, ArtifactPosition
 	return true;
 }
 
+bool CCallback::bulkMoveArtifacts(ObjectInstanceID srcHero, ObjectInstanceID dstHero, bool swap)
+{
+	BulkExchangeArtifacts bma(srcHero, dstHero, swap);
+	sendRequest(&bma);
+	return true;
```
Comment (alexvins) [inline CCallback.cpp:188]: Why return `true` always? Please declare method as `void`.

File: server/CQuery.cpp [hunk 5]
```diff
@@ -370,6 +370,9 @@ bool CGarrisonDialogQuery::blocksPack(const CPack * pack) const
 	}
 	if(auto dismiss = dynamic_ptr_cast<DisbandCreature>(pack))
 		return !vstd::contains(ourIds, dismiss->id);
+	
+	if (auto arts = dynamic_ptr_cast<BulkExchangeArtifacts>(pack))
+		return !vstd::contains(ourIds, arts->srcHero) || !vstd::contains(ourIds, arts->dstHero);
```
[the remark this answers is not in the document]
Comment (SoundSSGood) [reply server/CQuery.cpp:375]: CGarrisonDialogQuery class used also for hero exchange dialogs. As it noted in the comment in the class declaration. So yes, exchangingArmies id is same as hero id.

File: lib/CArtHandler.h [hunk 6]
```diff
@@ -363,4 +364,27 @@ class DLL_LINKAGE CArtifactSet
...
+
+protected:
+	ArtBearer::ArtBearer Bearer;
+};
+
+namespace ArtifactUtils
+{
+	// Calculates where an artifact gets placed when it gets transferred from one hero to another.
+	DLL_LINKAGE ArtifactPosition getArtifactDstPosition(const CArtifactInstance * artifact, const CArtifactSet * target,
+		ArtBearer::ArtBearer bearer);
```
Comment (nullkiller) [inline lib/CArtHandler.h:384]: Put each argument on a new line please

File: server/CGameHandler.cpp [hunk 7]
```diff
@@ -3916,10 +3916,96 @@ bool CGameHandler::moveArtifact(const ArtifactLocation &al1, const ArtifactLocat
...
+		moveArtsWorn(pdstHero, psrcHero, slotsDstSrc);
+		// Move over artifacts that are in backpack srcHero -> dstHero
+		moveArtsInBackpack(psrcHero, slotsSrcDst);
+		// Move over artifacts that are in backpack dstHero -> srcHero
+		moveArtsInBackpack(pdstHero, slotsDstSrc);
+	}
+	else
+	{
+		// Temporary fitting set for artifacts. Used to select available slots before sending data.
+		CArtifactFittingSet ArtFittingSet(pdstHero->bearerType());
```
Comment (nullkiller) [inline server/CGameHandler.cpp:3980]:
```suggestion
		CArtifactFittingSet artFittingSet(pdstHero->bearerType());
```

**Review by nullkiller — APPROVED** [review]

File: lib/CArtHandler.h (hunk 6 above)
Comment (SoundSSGood) [reply lib/CArtHandler.h:384]: Only in the header or in the c file too?

Comment (nullkiller) [reply lib/CArtHandler.h:384]: Both, standard practice to avoid long lines

**Review by Nordsoft91 — APPROVED** [review]

File: server/CGameHandler.cpp [hunk 8]
```diff
@@ -3916,10 +3916,96 @@ bool CGameHandler::moveArtifact(const ArtifactLocation &al1, const ArtifactLocat
...
+		{
+			for(auto & artifact : srcHero->artifactsWorn)
+			{
+				if(artifact.second.locked)
+					continue;
+				if(!ArtifactUtils::isArtRemovable(artifact))
+					continue;
+				slots.push_back(BulkMoveArtifacts::LinkedSlots(artifact.first, artifact.first));
+
+				if(ArtifactUtils::checkSpellbookIsNeeded(dstHero, artifact.second.getArt()->artType->id, artifact.first))
```
[the remark this answers is not in the document]
Comment (SoundSSGood) [reply server/CGameHandler.cpp]: fixed

**Pull request merged**